src/librustc/ty/layout.rs

```diff
@@ -833,7 +971,9 @@ pub enum Layout {
         nndiscr: u64,
         nonnull: Struct,
         // N.B. There is a 0 at the start, for LLVM GEP through a pointer.
-        discrfield: FieldPath
+        discrfield: FieldPath,
+        // Like discrfield, but in source order. For debuginfo.
+        discrfield_source: FieldPath
```

Comment:
It is very, very sad.

I don't know who mw is on here, so feel free to ping them.  I'm almost certain the conversation about this was one I had with them.  But the long and the short of it is that dwarf's equivalent of GEP is based off the source order, not the memory order.  This is then put into the name of the type as something like `$1$2` or somesuch (I don't remember the specific format offhand) so that the pretty printers for Rust can find the discriminant in this case and print the right variant.

Since layout doesn't keep around enough information to invert the path at that level, it's this or some sort of big function that duplicates a lot of what layout already does.  We compute the source-order path at this point, so I put it in.  If it is important to move this to debuginfo, it might be possible to do so.  Maybe reverting `non_zero_field_paths` and making the old version public, but adding a flag to avoid the translations?  This might work, but I'm not entirely convinced it's less sad.